File: src/builtin/stream.rs
use crate::builtin::buffer::{buffer_to_bytes, new_buffer_from_bytes};
use crate::value::Value;
use std::cell::RefCell;
use std::collections::HashMap;
use std::fs::{File, OpenOptions};
use std::io::{BufRead, BufReader, BufWriter, Read, Write};
use std::rc::Rc;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Mutex, OnceLock};

const READ_STREAM_ID_KEY: &str = "__read_stream_id";
const WRITE_STREAM_ID_KEY: &str = "__write_stream_id";

static READ_STREAM_SEQ: AtomicI64 = AtomicI64::new(1);
static WRITE_STREAM_SEQ: AtomicI64 = AtomicI64::new(1);
static READ_STREAMS: OnceLock<Mutex<HashMap<i64, BufReader<File>>>> = OnceLock::new();
static WRITE_STREAMS: OnceLock<Mutex<HashMap<i64, BufWriter<File>>>> = OnceLock::new();

fn readers() -> &'static Mutex<HashMap<i64, BufReader<File>>> {
    READ_STREAMS.get_or_init(|| Mutex::new(HashMap::new()))
}

fn writers() -> &'static Mutex<HashMap<i64, BufWriter<File>>> {
    WRITE_STREAMS.get_or_init(|| Mutex::new(HashMap::new()))
}

pub fn create_stream_module() -> Value {
    let mut ns = HashMap::new();
    insert_fn(&mut ns, "openRead", stream_open_read);
    insert_fn(&mut ns, "openWrite", stream_open_write);
    Value::Module(Rc::new(RefCell::new(ns)))
}

fn insert_fn(
    ns: &mut HashMap<String, Value>,
    name: &str,
    func: fn(&[Value]) -> Result<Value, String>,
) {
    ns.insert(
        name.to_string(),
        Value::NativeFunction {
            name: format!("stream::{}", name),
            func,
        },
    );
}

fn method(name: &str, func: fn(&[Value]) -> Result<Value, String>) -> Value {
    Value::NativeFunction {
        name: name.to_string(),
        func,
    }
}

fn as_string(v: &Value, fn_name: &str) -> Result<String, String> {
    match v {
        Value::String(s) => Ok(s.clone()),
        _ => Err(format!("{} expects string argument", fn_name)),
    }
}

fn as_bool(v: &Value, fn_name: &str) -> Result<bool, String> {
    match v {
        Value::Bool(b) => Ok(*b),
        _ => Err(format!("{} expects bool argument", fn_name)),
    }
}

fn read_stream_id_from_self(value: &Value) -> Result<i64, String> {
    let s = match value {
        Value::Struct(s) => s,
        _ => return Err("ReadStream method expects ReadStream object".to_string()),
    };
    let guard = s.borrow();
    match guard.get(READ_STREAM_ID_KEY) {
        Some(Value::Int(id)) => Ok(*id),
        _ => Err("Invalid ReadStream object".to_string()),
    }
}

fn write_stream_id_from_self(value: &Value) -> Result<i64, String> {
    let s = match value {
        Value::Struct(s) => s,
        _ => return Err("WriteStream method expects WriteStream object".to_string()),
    };
    let guard = s.borrow();
    match guard.get(WRITE_STREAM_ID_KEY) {
        Some(Value::Int(id)) => Ok(*id),
        _ => Err("Invalid WriteStream object".to_string()),
    }
}

fn build_read_stream(id: i64) -> Value {
    let mut fields = HashMap::new();
    fields.insert(READ_STREAM_ID_KEY.to_string(), Value::Int(id));
    fields.insert(
        "readBytes".to_string(),
        method("ReadStream::readBytes", read_stream_read_bytes),
    );
    fields.insert(
        "readLine".to_string(),
        method("ReadStream::readLine", read_stream_read_line),
    );
    fields.insert(
        "readAll".to_string(),
        method("ReadStream::readAll", read_stream_read_all),
    );
    fields.insert(
        "close".to_string(),
        method("ReadStream::close", read_stream_close),
    );
    Value::Struct(Rc::new(RefCell::new(fields)))
}

fn build_write_stream(id: i64) -> Value {
    let mut fields = HashMap::new();
    fields.insert(WRITE_STREAM_ID_KEY.to_string(), Value::Int(id));
    fields.insert(
        "writeBytes".to_string(),
        method("WriteStream::writeBytes", write_stream_write_bytes),
    );
    fields.insert(
        "writeText".to_string(),
        method("WriteStream::writeText", write_stream_write_text),
    );
    fields.insert(
        "flush".to_string(),
        method("WriteStream::flush", write_stream_flush),
    );
    fields.insert(
        "close".to_string(),
        method("WriteStream::close", write_stream_close),
    );
    Value::Struct(Rc::new(RefCell::new(fields)))
}

fn stream_open_read(args: &[Value]) -> Result<Value, String> {
    if args.len() != 2 {
        return Err("stream.openRead expects 1 argument (path)".to_string());
    }
    let path = as_string(&args[1], "stream.openRead")?;
    let file =
        File::open(&path).map_err(|e| format!("stream.openRead failed for '{}': {}", path, e))?;

    let id = READ_STREAM_SEQ.fetch_add(1, Ordering::Relaxed);
    let mut map = readers()
        .lock()
        .map_err(|_| "ReadStream store lock poisoned".to_string())?;
    map.insert(id, BufReader::new(file));
    Ok(build_read_stream(id))
}

fn stream_open_write(args: &[Value]) -> Result<Value, String> {
    if args.len() != 3 {
        return Err("stream.openWrite expects 2 arguments (path, append)".to_string());
    }
    let path = as_string(&args[1], "stream.openWrite")?;
    let append = as_bool(&args[2], "stream.openWrite")?;

    let mut opts = OpenOptions::new();
    opts.create(true).write(true);
    if append {
        opts.append(true);
    } else {
        opts.truncate(true);
    }

    let file = opts
        .open(&path)
        .map_err(|e| format!("stream.openWrite failed for '{}': {}", path, e))?;

    let id = WRITE_STREAM_SEQ.fetch_add(1, Ordering::Relaxed);
    let mut map = writers()
        .lock()
        .map_err(|_| "WriteStream store lock poisoned".to_string())?;
    map.insert(id, BufWriter::new(file));
    Ok(build_write_stream(id))
}

fn read_stream_read_bytes(args: &[Value]) -> Result<Value, String> {
    if args.len() != 2 {
        return Err("readStream.readBytes expects 1 argument (size)".to_string());
    }
    let id = read_stream_id_from_self(&args[0])?;
    let size = args[1].to_int()?;
    if size < 0 {
        return Err("readStream.readBytes expects non-negative size".to_string());
    }

    let mut map = readers()
        .lock()
        .map_err(|_| "ReadStream store lock poisoned".to_string())?;
    let reader = map
        .get_mut(&id)
        .ok_or_else(|| "ReadStream is closed or invalid".to_string())?;

    let mut buf = vec![0u8; size as usize];
    let n = reader
        .read(&mut buf)
        .map_err(|e| format!("readStream.readBytes failed: {}", e))?;
    if n == 0 {
        return Ok(Value::Null);
    }
    buf.truncate(n);
    Ok(new_buffer_from_bytes(buf))
}

fn read_stream_read_line(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("readStream.readLine expects no arguments".to_string());
    }
    let id = read_stream_id_from_self(&args[0])?;

    let mut map = readers()
        .lock()
        .map_err(|_| "ReadStream store lock poisoned".to_string())?;
    let reader = map
        .get_mut(&id)
        .ok_or_else(|| "ReadStream is closed or invalid".to_string())?;

    let mut bytes = Vec::new();
    let n = reader
        .read_until(b'\n', &mut bytes)
        .map_err(|e| format!("readStream.readLine failed: {}", e))?;
    if n == 0 {
        return Ok(Value::Null);
    }

    if bytes.ends_with(b"\n") {
        bytes.pop();
        if bytes.ends_with(b"\r") {
            bytes.pop();
        }
    }

    let line = String::from_utf8(bytes)
        .map_err(|e| format!("readStream.readLine invalid UTF-8: {}", e))?;
    Ok(Value::String(line))
}

fn read_stream_read_all(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("readStream.readAll expects no arguments".to_string());
    }
    let id = read_stream_id_from_self(&args[0])?;

    let mut map = readers()
        .lock()
        .map_err(|_| "ReadStream store lock poisoned".to_string())?;
    let reader = map
        .get_mut(&id)
        .ok_or_else(|| "ReadStream is closed or invalid".to_string())?;

    let mut out = Vec::new();
    reader
        .read_to_end(&mut out)
        .map_err(|e| format!("readStream.readAll failed: {}", e))?;
    Ok(new_buffer_from_bytes(out))
}

fn read_stream_close(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("readStream.close expects no arguments".to_string());
    }
    let id = read_stream_id_from_self(&args[0])?;
    let mut map = readers()
        .lock()
        .map_err(|_| "ReadStream store lock poisoned".to_string())?;
    map.remove(&id);
    Ok(Value::Null)
}

fn write_stream_write_bytes(args: &[Value]) -> Result<Value, String> {
    if args.len() != 2 {
        return Err("writeStream.writeBytes expects 1 argument (data)".to_string());
    }
    let id = write_stream_id_from_self(&args[0])?;
    let bytes = match &args[1] {
        Value::Struct(_) => buffer_to_bytes(&args[1])?,
        _ => return Err("writeStream.writeBytes expects Buffer".to_string()),
    };

    let mut map = writers()
        .lock()
        .map_err(|_| "WriteStream store lock poisoned".to_string())?;
    let writer = map
        .get_mut(&id)
        .ok_or_else(|| "WriteStream is closed or invalid".to_string())?;
    writer
        .write_all(&bytes)
        .map_err(|e| format!("writeStream.writeBytes failed: {}", e))?;
    Ok(Value::Null)
}

fn write_stream_write_text(args: &[Value]) -> Result<Value, String> {
    if args.len() != 2 {
        return Err("writeStream.writeText expects 1 argument (text)".to_string());
    }
    let id = write_stream_id_from_self(&args[0])?;
    let text = as_string(&args[1], "writeStream.writeText")?;

    let mut map = writers()
        .lock()
        .map_err(|_| "WriteStream store lock poisoned".to_string())?;
    let writer = map
        .get_mut(&id)
        .ok_or_else(|| "WriteStream is closed or invalid".to_string())?;
    writer
        .write_all(text.as_bytes())
        .map_err(|e| format!("writeStream.writeText failed: {}", e))?;
    Ok(Value::Null)
}

fn write_stream_flush(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("writeStream.flush expects no arguments".to_string());
    }
    let id = write_stream_id_from_self(&args[0])?;

    let mut map = writers()
        .lock()
        .map_err(|_| "WriteStream store lock poisoned".to_string())?;
    let writer = map
        .get_mut(&id)
        .ok_or_else(|| "WriteStream is closed or invalid".to_string())?;
    writer
        .flush()
        .map_err(|e| format!("writeStream.flush failed: {}", e))?;
    Ok(Value::Null)
}

fn write_stream_close(args: &[Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err("writeStream.close expects no arguments".to_string());
    }
    let id = write_stream_id_from_self(&args[0])?;

    let mut map = writers()
        .lock()
        .map_err(|_| "WriteStream store lock poisoned".to_string())?;
    if let Some(mut writer) = map.remove(&id) {
        writer
            .flush()
            .map_err(|e| format!("writeStream.close flush failed: {}", e))?;
    }
    Ok(Value::Null)
}

File: cxx/list/src/builtin/stream.cc
#include <builtin/stream.h>
#include <builtin/buffer.h>

#include <fstream>
#include <mutex>
#include <unordered_map>
#include <atomic>

namespace rumina {
namespace builtin {
namespace stream {

// using namespace rumina::builtin::buffer;

static std::atomic<int64_t> read_stream_seq{1};
static std::atomic<int64_t> write_stream_seq{1};
static std::mutex readers_mutex;
static std::mutex writers_mutex;
static std::unordered_map<int64_t, std::shared_ptr<std::ifstream>> readers;
static std::unordered_map<int64_t, std::shared_ptr<std::ofstream>> writers;

constexpr const char* READ_STREAM_ID_KEY = "__read_stream_id";
constexpr const char* WRITE_STREAM_ID_KEY = "__write_stream_id";

// 创建流模块
Value create_stream_module() {
    auto ns = std::make_shared<std::unordered_map<std::string, Value>>();
    
    (*ns)["openRead"] = Value::makeNativeFunction("stream::openRead", stream_open_read);
    (*ns)["openWrite"] = Value::makeNativeFunction("stream::openWrite", stream_open_write);
    
    return Value::makeModule(ns);
}

// 创建方法
static Value method(const std::string& name, NativeFunction func) {
    return Value::makeNativeFunction(name, func);
}

// 获取字符串参数
static std::string as_string(const Value& v, const std::string& fn_name) {
    if (v.getType() != Value::Type::String) {
        throw std::runtime_error(fn_name + " expects string argument");
    }
    return v.getString();
}

// 获取布尔参数
static bool as_bool(const Value& v, const std::string& fn_name) {
    if (v.getType() != Value::Type::Bool) {
        throw std::runtime_error(fn_name + " expects bool argument");
    }
    return v.getBool();
}

// 从ReadStream对象获取ID
static int64_t read_stream_id_from_self(const Value& value) {
    if (value.getType() != Value::Type::Struct) {
        throw std::runtime_error("ReadStream method expects ReadStream object");
    }
    
    auto fields = value.getStruct();
    auto it = fields->find(READ_STREAM_ID_KEY);
    if (it == fields->end() || it->second.getType() != Value::Type::Int) {
        throw std::runtime_error("Invalid ReadStream object");
    }
    
    return it->second.getInt();
}

// 从WriteStream对象获取ID
static int64_t write_stream_id_from_self(const Value& value) {
    if (value.getType() != Value::Type::Struct) {
        throw std::runtime_error("WriteStream method expects WriteStream object");
    }
    
    auto fields = value.getStruct();
    auto it = fields->find(WRITE_STREAM_ID_KEY);
    if (it == fields->end() || it->second.getType() != Value::Type::Int) {
        throw std::runtime_error("Invalid WriteStream object");
    }
    
    return it->second.getInt();
}

// 构建ReadStream对象
static Value build_read_stream(int64_t id) {
    auto fields = std::make_shared<std::unordered_map<std::string, Value>>();
    
    (*fields)[READ_STREAM_ID_KEY] = Value(id);
    (*fields)["readBytes"] = method("ReadStream::readBytes", read_stream_read_bytes);
    (*fields)["readUntil"] = method("ReadStream::readUntil", read_stream_read_until);
    (*fields)["readAll"] = method("ReadStream::readAll", read_stream_read_all);
    (*fields)["seek"] = method("ReadStream::seek", read_stream_seek);
    (*fields)["tell"] = method("ReadStream::tell", read_stream_tell);
    (*fields)["isClosed"] = method("ReadStream::isClosed", read_stream_is_closed);
    (*fields)["close"] = method("ReadStream::close", read_stream_close);
    
    return Value::makeStruct(fields);
}

// 构建WriteStream对象
static Value build_write_stream(int64_t id) {
    auto fields = std::make_shared<std::unordered_map<std::string, Value>>();
    
    (*fields)[WRITE_STREAM_ID_KEY] = Value(id);
    (*fields)["writeBytes"] = method("WriteStream::writeBytes", write_stream_write_bytes);
    (*fields)["writeText"] = method("WriteStream::writeText", write_stream_write_text);
    (*fields)["flush"] = method("WriteStream::flush", write_stream_flush);
    (*fields)["seek"] = method("WriteStream::seek", write_stream_seek);
    (*fields)["tell"] = method("WriteStream::tell", write_stream_tell);
    (*fields)["isClosed"] = method("WriteStream::isClosed", write_stream_is_closed);
    (*fields)["close"] = method("WriteStream::close", write_stream_close);
    
    return Value::makeStruct(fields);
}

// stream.openRead(path)
Value stream_open_read(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("stream.openRead expects 1 argument (path)");
    }
    
    std::string path = as_string(args[0], "stream.openRead");
    
    auto file = std::make_shared<std::ifstream>(path, std::ios::binary);
    if (!file->is_open()) {
        throw std::runtime_error("stream.openRead failed for '" + path + "'");
    }
    
    int64_t id = read_stream_seq.fetch_add(1);
    
    {
        std::lock_guard<std::mutex> lock(readers_mutex);
        readers[id] = file;
    }
    
    return build_read_stream(id);
}

// stream.openWrite(path, append)
Value stream_open_write(const std::vector<Value>& args) {
    if (args.size() != 2) {
        throw std::runtime_error("stream.openWrite expects 2 arguments (path, append)");
    }
    
    std::string path = as_string(args[0], "stream.openWrite");
    bool append = as_bool(args[1], "stream.openWrite");
    
    std::ios_base::openmode mode = std::ios::binary;
    if (append) {
        mode |= std::ios::app;
    } else {
        mode |= std::ios::trunc;
    }
    
    auto file = std::make_shared<std::ofstream>(path, mode);
    if (!file->is_open()) {
        throw std::runtime_error("stream.openWrite failed for '" + path + "'");
    }
    
    int64_t id = write_stream_seq.fetch_add(1);
    
    {
        std::lock_guard<std::mutex> lock(writers_mutex);
        writers[id] = file;
    }
    
    return build_write_stream(id);
}

// 获取分隔符字节
static std::vector<uint8_t> delimiter_bytes(const Value& value) {
    if (value.getType() == Value::Type::String) {
        const std::string& s = value.getString();
        return std::vector<uint8_t>(s.begin(), s.end());
    } else if (value.getType() == Value::Type::Struct) {
        return buffer::buffer_to_bytes(value);
    } else {
        throw std::runtime_error("readStream.readUntil expects delimiter as String or Buffer");
    }
}

// readStream.readBytes(size)
Value read_stream_read_bytes(const std::vector<Value>& args) {
    if (args.size() != 2) {
        throw std::runtime_error("readStream.readBytes expects 1 argument (size)");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    int64_t size = args[1].toInt();
    if (size < 0) {
        throw std::runtime_error("readStream.readBytes expects non-negative size");
    }
    
    std::shared_ptr<std::ifstream> file;
    {
        std::lock_guard<std::mutex> lock(readers_mutex);
        auto it = readers.find(id);
        if (it == readers.end()) {
            throw std::runtime_error("ReadStream is closed or invalid");
        }
        file = it->second;
    }
    
    std::vector<uint8_t> buf(static_cast<size_t>(size));
    file->read(reinterpret_cast<char*>(buf.data()), size);
    size_t n = file->gcount();
    
    if (n == 0) {
        return Value();
    }
    
    buf.resize(n);
    return buffer::new_buffer_from_bytes(buf);
}

// readStream.readUntil(delimiter, maxBytes?)
Value read_stream_read_until(const std::vector<Value>& args) {
    if (args.size() < 2 || args.size() > 3) {
        throw std::runtime_error("readStream.readUntil expects 1 or 2 arguments (delimiter, maxBytes?)");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    std::vector<uint8_t> delim = delimiter_bytes(args[1]);
    
    if (delim.empty()) {
        throw std::runtime_error("readStream.readUntil delimiter cannot be empty");
    }
    
    std::optional<size_t> max_bytes;
    if (args.size() == 3) {
        int64_t mb = args[2].toInt();
        if (mb < 0) {
            throw std::runtime_error("readStream.readUntil maxBytes must be non-negative");
        }
        max_bytes = static_cast<size_t>(mb);
    }
    
    std::shared_ptr<std::ifstream> file;
    {
        std::lock_guard<std::mutex> lock(readers_mutex);
        auto it = readers.find(id);
        if (it == readers.end()) {
            throw std::runtime_error("ReadStream is closed or invalid");
        }
        file = it->second;
    }
    
    std::vector<uint8_t> out;
    bool reached_eof = false;
    
    while (true) {
        if (max_bytes.has_value() && out.size() >= max_bytes.value()) {
            break;
        }
        
        char c;
        if (!file->get(c)) {
            reached_eof = true;
            break;
        }
        
        out.push_back(static_cast<uint8_t>(c));
        
        if (out.size() >= delim.size()) {
            bool match = true;
            for (size_t i = 0; i < delim.size(); ++i) {
                if (out[out.size() - delim.size() + i] != delim[i]) {
                    match = false;
                    break;
                }
            }
            
            if (match) {
                out.resize(out.size() - delim.size());
                break;
            }
        }
    }
    
    if (out.empty() && reached_eof) {
        return Value();
    }
    
    return buffer::new_buffer_from_bytes(out);
}

// readStream.readAll()
Value read_stream_read_all(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("readStream.readAll expects no arguments");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    
    std::shared_ptr<std::ifstream> file;
    {
        std::lock_guard<std::mutex> lock(readers_mutex);
        auto it = readers.find(id);
        if (it == readers.end()) {
            throw std::runtime_error("ReadStream is closed or invalid");
        }
        file = it->second;
    }
    
    std::vector<uint8_t> out;
    char c;
    while (file->get(c)) {
        out.push_back(static_cast<uint8_t>(c));
    }
    
    return buffer::new_buffer_from_bytes(out);
}

// readStream.seek(offset)
Value read_stream_seek(const std::vector<Value>& args) {
    if (args.size() != 2) {
        throw std::runtime_error("readStream.seek expects 1 argument (offset)");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    int64_t offset = args[1].toInt();
    if (offset < 0) {
        throw std::runtime_error("readStream.seek expects non-negative offset");
    }
    
    std::shared_ptr<std::ifstream> file;
    {
        std::lock_guard<std::mutex> lock(readers_mutex);
        auto it = readers.find(id);
        if (it == readers.end()) {
            throw std::runtime_error("ReadStream is closed or invalid");
        }
        file = it->second;
    }
    
    file->seekg(static_cast<std::streamoff>(offset), std::ios::beg);
    if (file->fail()) {
        throw std::runtime_error("readStream.seek failed");
    }
    
    std::streampos pos = file->tellg();
    return Value(static_cast<int64_t>(pos));
}

// readStream.tell()
Value read_stream_tell(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("readStream.tell expects no arguments");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    
    std::shared_ptr<std::ifstream> file;
    {
        std::lock_guard<std::mutex> lock(readers_mutex);
        auto it = readers.find(id);
        if (it == readers.end()) {
            throw std::runtime_error("ReadStream is closed or invalid");
        }
        file = it->second;
    }
    
    std::streampos pos = file->tellg();
    return Value(static_cast<int64_t>(pos));
}

// readStream.isClosed()
Value read_stream_is_closed(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("readStream.isClosed expects no arguments");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    
    std::lock_guard<std::mutex> lock(readers_mutex);
    return Value(readers.find(id) == readers.end());
}

// readStream.close()
Value read_stream_close(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("readStream.close expects no arguments");
    }
    
    int64_t id = read_stream_id_from_self(args[0]);
    
    std::lock_guard<std::mutex> lock(readers_mutex);
    readers.erase(id);
    
    return Value();
}

// writeStream.writeBytes(data)
Value write_stream_write_bytes(const std::vector<Value>& args) {
    if (args.size() != 2) {
        throw std::runtime_error("writeStream.writeBytes expects 1 argument (data)");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    std::vector<uint8_t> bytes = buffer::buffer_to_bytes(args[1]);
    
    std::shared_ptr<std::ofstream> file;
    {
        std::lock_guard<std::mutex> lock(writers_mutex);
        auto it = writers.find(id);
        if (it == writers.end()) {
            throw std::runtime_error("WriteStream is closed or invalid");
        }
        file = it->second;
    }
    
    file->write(reinterpret_cast<const char*>(bytes.data()), bytes.size());
    if (file->fail()) {
        throw std::runtime_error("writeStream.writeBytes failed");
    }
    
    return Value();
}

// writeStream.writeText(text)
Value write_stream_write_text(const std::vector<Value>& args) {
    if (args.size() != 2) {
        throw std::runtime_error("writeStream.writeText expects 1 argument (text)");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    std::string text = as_string(args[1], "writeStream.writeText");
    
    std::shared_ptr<std::ofstream> file;
    {
        std::lock_guard<std::mutex> lock(writers_mutex);
        auto it = writers.find(id);
        if (it == writers.end()) {
            throw std::runtime_error("WriteStream is closed or invalid");
        }
        file = it->second;
    }
    
    file->write(text.data(), text.size());
    if (file->fail()) {
        throw std::runtime_error("writeStream.writeText failed");
    }
    
    return Value();
}

// writeStream.flush()
Value write_stream_flush(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("writeStream.flush expects no arguments");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    
    std::shared_ptr<std::ofstream> file;
    {
        std::lock_guard<std::mutex> lock(writers_mutex);
        auto it = writers.find(id);
        if (it == writers.end()) {
            throw std::runtime_error("WriteStream is closed or invalid");
        }
        file = it->second;
    }
    
    file->flush();
    if (file->fail()) {
        throw std::runtime_error("writeStream.flush failed");
    }
    
    return Value();
}

// writeStream.seek(offset)
Value write_stream_seek(const std::vector<Value>& args) {
    if (args.size() != 2) {
        throw std::runtime_error("writeStream.seek expects 1 argument (offset)");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    int64_t offset = args[1].toInt();
    if (offset < 0) {
        throw std::runtime_error("writeStream.seek expects non-negative offset");
    }
    
    std::shared_ptr<std::ofstream> file;
    {
        std::lock_guard<std::mutex> lock(writers_mutex);
        auto it = writers.find(id);
        if (it == writers.end()) {
            throw std::runtime_error("WriteStream is closed or invalid");
        }
        file = it->second;
    }
    
    file->seekp(static_cast<std::streamoff>(offset), std::ios::beg);
    if (file->fail()) {
        throw std::runtime_error("writeStream.seek failed");
    }
    
    std::streampos pos = file->tellp();
    return Value(static_cast<int64_t>(pos));
}

// writeStream.tell()
Value write_stream_tell(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("writeStream.tell expects no arguments");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    
    std::shared_ptr<std::ofstream> file;
    {
        std::lock_guard<std::mutex> lock(writers_mutex);
        auto it = writers.find(id);
        if (it == writers.end()) {
            throw std::runtime_error("WriteStream is closed or invalid");
        }
        file = it->second;
    }
    
    std::streampos pos = file->tellp();
    return Value(static_cast<int64_t>(pos));
}

// writeStream.isClosed()
Value write_stream_is_closed(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("writeStream.isClosed expects no arguments");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    
    std::lock_guard<std::mutex> lock(writers_mutex);
    return Value(writers.find(id) == writers.end());
}

// writeStream.close()
Value write_stream_close(const std::vector<Value>& args) {
    if (args.size() != 1) {
        throw std::runtime_error("writeStream.close expects no arguments");
    }
    
    int64_t id = write_stream_id_from_self(args[0]);
    
    std::lock_guard<std::mutex> lock(writers_mutex);
    auto it = writers.find(id);
    if (it != writers.end()) {
        it->second->flush();
        writers.erase(it);
    }
    
    return Value();
}

} // namespace stream
} // namespace builtin
} // namespace rumina
